Missing string in admin mod_menu

infograf768 opened the pull request:
### Summary of Changes
The string
`MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE=""`

should NOT have been deleted! ( see https://github.com/joomla/joomla-cms/pull/12467 )
When the value is not set it automatically loads the Languages forums page with the correct text
For en-GB it gives

![screen shot 2017-09-02 at 11 58 10](https://user-images.githubusercontent.com/869724/29994701-165e4178-8fd6-11e7-95d1-f0c0e79c6f4f.png)

When value is set to the correct language forum (for example `19` for French, it loads the French Language forum.

![screen shot 2017-09-02 at 11 59 31](https://user-images.githubusercontent.com/869724/29994711-3bf62d2e-8fd6-11e7-9481-0f91ab0aa6bf.png)

infograf768 committed:
Missing string in admin mod_menu

ghost commented:
Got same Menu as shown in above Screenshot for French Lang. without Pull Request:
![bildschirmfoto 2017-09-02 um 12 08 28](https://user-images.githubusercontent.com/8235763/29994767-7abf4c24-8fd7-11e7-8a8b-6f5e4e0d5dcb.jpg)

infograf768 commented:
@franz-wohlkoenig 
yes you get that because we kept the string in fr-FR although it is deleted in en-GB. That is how I found it was missing in en-GB.
It is necessary to have the string as it shows to Translation Teams. For example TTs using crowdin will not even know it exists...

remark the comment above the missing string in en-GB.mod_menu.ini:
`; The string below will be used if MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE has a value, i.e the # of the specific language forum in https://forum.joomla.org/. Use something like 'Official english forum'.`
But no string below...

infograf768 commented:
Can be merged on review.

wojsmol commented:
@infograf768 IMHO this PR is wrong see code 
https://github.com/joomla/joomla-cms/blob/2ba24ebba5de653d75aaf8ecefc422e199108ebe/administrator/components/com_menus/presets/joomla.xml#L480-L487
and
https://github.com/joomla/joomla-cms/blob/2ba24ebba5de653d75aaf8ecefc422e199108ebe/administrator/components/com_admin/views/help/tmpl/langforum.php#L12-L13

infograf768 commented:
@wojsmol 
That is real weird!!!. 
This has been changed some time ago without my knowledge.
It was still working in 3.7.4 !!!
I agree we have an issue here.

@mbabker 
I will look into this tomorrow. mod_menu.ini is always loaded. Will see why it was changed in 3.8.0

wojsmol commented:
@infograf768 This has been changed as part of https://github.com/joomla/joomla-cms/pull/16451.

infograf768 commented:
@wojsmol 
Yes, I found that too.
I am trying now to get it right and if we can't use the mod_menu string, I will inform TTs of this change.

wojsmol commented:
Another good idea would be to replace the link to the subforum in forum.joomla.org with a link to an external community forum. This would be useful if the .org subforum is less active then the external forum.

infograf768 commented:
The new custom admin menus have also broken the mod_menu parameter `forum_url`

infograf768 commented:
> Another good idea would be to replace the link to the subforum in forum.joomla.org with a link to an external community forum.

We never allowed that as it would allow Translators to give an advantage to their own site. It's part of the policy with Translation Teams.

The mod_menu `forum_url` parameter was created to let custom distribs add such a link, or users to add or modify this. Not the TTs.

infograf768 commented:
Before the custom admin menus and presets we had in `enabled`

```
if ($forum_url = $params->get('forum_url'))
	{
		$menu->addChild(
			new JMenuNode(JText::_('MOD_MENU_HELP_SUPPORT_CUSTOM_FORUM'), $forum_url, 'class:help-forum', false, '_blank')
		);
	}
	$debug = $lang->setDebug(false);
	if ($lang->hasKey('MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE') && JText::_('MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE') != '')
	{
		$forum_url = 'http://forum.joomla.org/viewforum.php?f=' . (int) JText::_('MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE');
		$lang->setDebug($debug);
		$menu->addChild(
			new JMenuNode(JText::_('MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM'), $forum_url, 'class:help-forum', false, '_blank')
		);
	}
```

infograf768 commented:
@izharaazmi 
Please look at that urgently.

It looks like we can't load mod_menu strings in langforum.php.
If we could manage to do that, we could at least not modify that part of the strings.

We would just have to modify langforum.php this way
```
$forumId   = (int) JText::_('MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE');

if (empty($forumId))
{
	$forumId = 511;
}

$forum_url = 'https://forum.joomla.org/viewforum.php?f=' . $forumId;

JFactory::getApplication()->redirect($forum_url);
```
and take off
`COM_ADMIN_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE="511"`

Otherwise, we will revert this PR, then make a new PR where the comments in mod_menu.ini are modified and add a comment in com_admin.ini new strings, then inform TTs.

We also need to correct the problem with the mod_menu `forum_url` parameter.

Still working on it though.

izharaazmi commented:
@infograf768 I'm traveling right now. I'll see to it tomorrow when at desk.

infograf768 commented:
For the custom lang forum, keeping this PR, we can do in helpforum.php

```
<?php
/**
 * @package     Joomla.Administrator
 * @subpackage  com_admin
 *
 * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see LICENSE.txt
 */

defined('_JEXEC') or die;

JFactory::getLanguage()->load('mod_menu', JPATH_ADMINISTRATOR, null, false, true);

$forumId   = (int) JText::_('MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE');

if (empty($forumId))
{
	$forumId = 511;
}

$forum_url = 'https://forum.joomla.org/viewforum.php?f=' . $forumId;

JFactory::getApplication()->redirect($forum_url);

```
and take off the com_admin string
This will load the mod_menu string correctly.

@wojsmol 
Can you install the French language and make this modification?
Then use French as backend language.
It should now load the French forum directly as we have the string in fr-FR

For the mod_menu param, it is another story.

wojsmol commented:
@infograf768 This works as described.

izharaazmi commented:
@infograf768 sorry that I don't understand what we are trying to achieve here. Can you please summarise it for me. Thank you.

infograf768 commented:
@wojsmol 
Thanks for testing.

@izharaazmi 
I started working on the custom forum url depending on the module parameter.
To do this I added a `customforum.php` in `com_admin/views/help/`
and modified the default presets.
One issue remains: How to prevent displaying that submenu when the module parameter is empty.

See:
[help.diff.zip](https://github.com/joomla/joomla-cms/files/1272862/help.diff.zip)

infograf768 commented:
@izharaazmi 
siesta now. will explain all later

izharaazmi commented:
Yeah right. I'll think about that and come up with a solution hopefully.

infograf768 commented:
@izharaazmi 

We have 2 issues.

1. When you created the `langforum.php` and the new string `COM_ADMIN_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE` you did not remark (although you deleted it) that the string `MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE` already existed and was used by TTs to link to their forum.joomla.org specific language forum.
That was clearly explained in the mod_menu.ini file (see above).

The patch I propose above https://github.com/joomla/joomla-cms/pull/17836#issuecomment-326788375 solves that issue. 
There is no reason to create a new string as we have an already existing one. Was only necessary to load the mod_menu.ini file.
Setting the string to "" in en-GB (done with this already merged PR) is covered by the patch if the string remains unchanged by some TTs.

The comment in the mod_menu.ini file has also to be modified:
`; The string below will be used if MOD_MENU_HELP_SUPPORT_OFFICIAL_LANGUAGE_FORUM_VALUE has a value, i.e the # of the specific language forum in https://forum.joomla.org/. Use something like 'Official english forum'.` 
as the logic has changed, i.e. one now always get that link, whether it is to the specific language forum (19 for French) or the part of the forum (511) displaying all language forums.

2. When implementing custom admin menus, you forgot that mod_menu was proposing a specific parameter to add a submenu in the help forum to anything one wanted, more specifically for Language Communities to add a link to their own forum.

![screen shot 2017-09-03 at 16 17 45](https://user-images.githubusercontent.com/869724/30003784-7d26633e-90c3-11e7-84ec-333b05382c59.png)

This was done via the code in `.../administrator/modules/mod_menu/tmpl/default_enabled.php`

```
if ($forum_url = $params->get('forum_url'))
	{
		$menu->addChild(
			new JMenuNode(JText::_('MOD_MENU_HELP_SUPPORT_CUSTOM_FORUM'), $forum_url, 'class:help-forum', false, '_blank')
		);
	}
```
That link may also be added or suggested by 3pd extensions. 
For example it can be `https://www.akeebabackup.com/documentation.html`

This is what I started working on in the .zip diff I posted above and where I could not find a simple way to not display the Custom Forum submenu when the field is empty.

in mod_menu.ini, we have
`; the string below will be used if the localised sample data contains a URL for the desired community forum or if the 'Custom Support Forum' field parameter in the Administrator Menu module contains a URL`

`MOD_MENU_HELP_SUPPORT_CUSTOM_FORUM="Custom Support Forum"`

Hope this helps understand.

izharaazmi commented:
Thank you @infograf768, for the details and now I understand it clear.

I recently came across 'com_pagebuilder' and another component (can't recall the name) that injected its menu in admin menu prior to J3.7 admin menu manager. Looks like now its not possible (or not easy) to do that anymore.

So I was wondering whether we should allow such things or not. If yes, then should there be a plugin event trigger that can allow us to manipulate menu items via plugin logic.

If we choose to go that way, the same could be useful for this too. Otherwise, I'd try to find something better. Please advise.

infograf768 commented:
@izharaazmi 
For the moment, I will make a PR to solve the first issue as we must get this in 3.8.0

For the second issue, can't you devise a way to display such a conditional in your code to allow displaying or not one of the `<menuitem` in the presets?

In my zip above, I added
```
<menuitem
			type="url"
			target="_blank"
			title="MOD_MENU_HELP_SUPPORT_CUSTOM_FORUM"
			link="index.php?option=com_admin&amp;view=help&amp;layout=customforum"
			class="class:help-forum"
			scope="help"
		/>
```
and the conditional is the same as in the proposed new customforum.php
```
$module = JModuleHelper::getModule('mod_menu');
$params = new JRegistry($module->params);

if ($params->get('forum_url'))
etc.
```

izharaazmi commented:
Sure, I must get that work.

infograf768 commented:
Please test https://github.com/joomla/joomla-cms/pull/17857
@wojsmol 
@izharaazmi 
as we have to get this urgently in staging

infograf768 commented:
@izharaazmi 
How does it go with the second issue? It is a regression.

izharaazmi commented:
I've solved the hiding item conditionally issue. However, I'd be able to do
the PR on Friday only.

infograf768 commented:
👍
